File: ej3/ej3.pl
% # -- Resolver un problema para una librería
% # -- ¿Es cierto que alguien escribió una determinada obra?
% # -- ¿Quién o quienes escribieron una obra?
% # -- ¿Qué obra escribió cierta persona?
% # -- Si es cierto que cierta persona escribió alguna obra, sin importar cual.
% # -- Si es cierto que cierta obra existe.

% # -- Base de DATOS
% obra(Titulo, Autores)
titulo(sandman)
titulo(socorro)
titulo(buenosPresagios)

autor(gaiman)
autor(elsaBornemann)
autor(pratchett)

% Relaciones donde se pueden hacer todas las consultas
obra(sandman, gaiman).
obra(socorro, elsaBornemann).
obra(buenosPresagios, gaiman).
obra(buenosPresagios, pratchett).

% Consultas
obra(sandman,coscu)
False

obra(buenosPresagios, Autores)
    gaiman
    prachett

obra(Obra, gaiman)
    buenosPresagios
    sandman

obra(_, gaiman)
    True

titulo(coscu)
    False

File: ej3/ej3.hs
-- Resolver un problema para una librería
-- Si es cierto que cierta persona escribió alguna obra, sin importar cual.
-- Si es cierto que cierta obra existe.
-- Resolver un problema para una librería
type Titulo = String
type Autor = String
data Obra = UnaObra {titulo:: Titulo, autores:: [Autor]}

-- Base de DATOS
todasLasObras::[Obra]
todasLasObras = [UnaObra “sandman” [“gaiman”], 
                UnaObra “socorro” [“ElsaBornemann”], 
                UnaObra “buenos presagios” [“gaiman”, “pratchett”]]

-- ¿Es cierto que alguien escribió una determinada obra?
alguienEscribio :: Titulo -> Bool
alguienEscribio unTitulo = any (\obra -> titulo obra == unTitulo) todasLasObras)
-- en caso de que se pueda tener una lista vacia de autores e igualmente exista la obra
alguienEscribio :: Titulo -> Bool
alguienEscribio unTitulo = any (\obra -> titulo obra == unTitulo && not (null (autores obra))) todasLasObras

-- ¿Quién o quienes escribieron una obra?
quienesEscribieron :: Obra -> [String]
quienesEscribieron obra = autores obra

-- ¿Qué obras escribió cierta persona?
queObrasEscribio :: Autor -> [Obra]
queObrasEscribio unAutor = filter(\obra-> any(\autor -> autor == unAutor) (autores obra) ) todasLasObras